File: algorithm/프로그래머스/2/176962. 과제 진행하기/과제 진행하기.go
import (
	"sort"
	"strconv"
	"strings"
)

type Plan struct {
    name string
    start int
    playtime int
}

func solution(plans [][]string) []string {
    answer := make([]string, 0, len(plans))
    sort.Slice(plans, func(i, j int) bool {return plans[i][1] < plans[j][1]})
    
    newPlan := make([]Plan, len(plans))
    for i, plan := range plans {
        startString := strings.Split(plan[1], ":")
        h, _ := strconv.Atoi(startString[0])
        m, _ := strconv.Atoi(startString[1])
        p, _ := strconv.Atoi(plan[2])
        
        newPlan[i] = Plan{
            name: plan[0],
            start: 60 * h + m,
            playtime: p,
        }
    }
    
    var q []Plan
    for i, plan := range newPlan {
        if i == len(newPlan) - 1 {
            answer = append(answer, plan.name)
            break
        }
        
        nextPlan := newPlan[i+1]
        if plan.start + plan.playtime <= nextPlan.start {
            answer = append(answer, plan.name)
            
            remainTime := nextPlan.start - (plan.start + plan.playtime)
            for len(q) > 0 && remainTime > 0 {
                holdPlan := q[len(q) - 1]
                q = q[:len(q) - 1]
                
                if holdPlan.playtime <= remainTime {
                    answer = append(answer, holdPlan.name)
                    remainTime -= holdPlan.playtime
                } else {
                    holdPlan.playtime = holdPlan.playtime - remainTime
                    remainTime = 0
                    q = append(q, holdPlan)
                }
            }
        } else {
            q = append(q, Plan{
                name: plan.name,
                start: plan.start,
                playtime: plan.start + plan.playtime - nextPlan.start,
            })
        }
    }
    
    for len(q) > 0 {
        holdPlan := q[len(q) - 1]
        q = q[:len(q) - 1]
        answer = append(answer, holdPlan.name)
    }
    
    return answer
}

File: algorithm/프로그래머스/2/176962. 과제 진행하기/과제 진행하기.py
def solution(plans):
    answer = []
    
    sorted_plans = sorted(plans, key=lambda x: x[1])
    for i, plan in enumerate(sorted_plans):
        h, m = plan[1].split(":")
        sorted_plans[i][1] = 60 * int(h) + int(m)
        sorted_plans[i][2] = int(plan[2])
        
    holdings = []
    for i, plan in enumerate(sorted_plans):
        if i == len(sorted_plans) - 1:
            answer.append(plan[0])
            break
            
        next_plan = sorted_plans[i+1]
        if plan[1] + plan[2] <= next_plan[1]:
            answer.append(plan[0])
            remain_time = next_plan[1] - (plan[1] + plan[2])
            while holdings and remain_time > 0:
                hold_plan = holdings.pop()
                if hold_plan[2] <= remain_time:
                    answer.append(hold_plan[0])
                    remain_time -= hold_plan[2]
                else:
                    hold_plan = [hold_plan[0], hold_plan[1], hold_plan[2] - remain_time]
                    remain_time = 0
                    holdings.append(hold_plan)
        else:
            holdings.append([plan[0], plan[1], plan[1] + plan[2] - next_plan[1]])        
                
    while holdings:
        answer.append(holdings.pop()[0])
    
    return answer